libvast/src/table_slice.cpp

```diff
@@ -823,4 +815,266 @@ auto resolve_operand(const table_slice& slice, const operand& op)
+  -> std::shared_ptr<arrow::Array> {
+  VAST_ASSERT(not list_offsets.empty());
+  auto it = list_offsets.begin();
+  auto result = *it++;
+  auto builder = arrow::Int32Builder{};
+  for (; it < list_offsets.end(); ++it) {
+    for (const auto& index : static_cast<const arrow::Int32Array&>(*result)) {
+      VAST_ASSERT(index);
+      const auto& next = static_cast<const arrow::Int32Array&>(**it);
+      VAST_ASSERT(not next.IsNull(*index));
+      auto append_result = builder.Append(next.Value(*index));
+      VAST_ASSERT(append_result.ok(), append_result.ToString().c_str());
+    }
+    result = builder.Finish().ValueOrDie();
+  }
+  return result;
+}
+
+auto make_flatten_transformation(
+  std::string_view separator, const std::string& name_prefix,
+  std::vector<std::shared_ptr<arrow::Array>> list_offsets)
+  -> indexed_transformation::function_type;
+
+auto flatten_record(
+  std::string_view separator, std::string_view name_prefix,
+  const std::vector<std::shared_ptr<arrow::Array>>& list_offsets,
+  struct record_type::field field, const std::shared_ptr<arrow::Array>& array)
+  -> std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+
+auto flatten_list(std::string_view separator, std::string_view name_prefix,
+                  std::vector<std::shared_ptr<arrow::Array>> list_offsets,
+                  struct record_type::field field,
+                  const std::shared_ptr<arrow::Array>& array)
+  -> std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+  using result_type = std::vector<
+    std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+  const auto& lt = caf::get<list_type>(field.type);
+  auto list_array
+    = std::static_pointer_cast<type_to_arrow_array_t<list_type>>(array);
+  list_offsets.push_back(list_array->offsets());
+  auto f = detail::overload{
+    [&]<concrete_type Type>(const Type&) -> result_type {
+      auto result = result_type{};
+      if (list_offsets.empty()) {
+        result.push_back({
+          {
+            fmt::format("{}{}", name_prefix, field.name),
+            field.type,
+          },
+          array,
+        });
+      } else {
+        auto combined_offsets = combine_offsets(list_offsets);
+        result.push_back({
+          {
+            fmt::format("{}{}", name_prefix, field.name),
+            field.type,
+          },
+          arrow::ListArray::FromArrays(*combined_offsets, *list_array->values())
+            .ValueOrDie(),
+        });
+      }
+      return result;
+    },
+    [&](const list_type& lt) -> result_type {
+      return flatten_list(separator, name_prefix, std::move(list_offsets),
+                          {field.name, lt}, list_array->values());
+    },
+    [&](const record_type& rt) -> result_type {
+      return flatten_record(separator, name_prefix, list_offsets,
+                            {field.name, rt}, list_array->values());
```

Comment:
`flatten_record` must pass the list offsets to every field, so it cannot move them. `flatten_list` only passed the list offsets to its inner values array, so it can always move them. So this is intended.